### PR TITLE
Add support for BeDbUri

### PR DESCRIPTION
**Pull request by khanaffan**

### CODE REVIEW

File: iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h [hunk 1]
```diff
@@ -3310,6 +3330,205 @@ struct EXPORT_VTABLE_ATTRIBUTE Db : NonCopyableClass
...
+        constexpr static const char* PARAM_TxnMode = "txn_mode";
+        constexpr static const char* PARAM_Vfs = "vfs";
+        BE_SQLITE_EXPORT static bool IsKnownParameter(Utf8CP str);
+        BE_SQLITE_EXPORT static std::set<Utf8CP, NoCaseCompare> const& GetKnownParameters();
+        BE_SQLITE_EXPORT static Db::OpenMode ParseOpenMode(Utf8StringCR str);
+        BE_SQLITE_EXPORT static Utf8String ToString(Db::OpenMode);
+    };
+
+private:
+    Utf8String _canonicalPath;
```
Comment (kabentley) [inline iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h:3399]: use "m_" for all member names

**Review by kabentley — CHANGES_REQUESTED** [review]

File: iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h (hunk 1 above)
Comment (khanaffan) [reply iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h]: Fixed.

File: iModelCore/BeSQLite/BeSQLite.cpp [hunk 2]
```diff
@@ -6422,3 +6465,202 @@ DbResult Db::Deserialize(DbBuffer& buffer, DbR db, DbDeserializeOptions opts, co
     }
     return rc;
 }
+
+Utf8String BeDbUri::Normalize(Utf8StringCR messy_path) {
+    std::filesystem::path path((std::string)messy_path);
+    std::filesystem::path canonicalPath = std::filesystem::weakly_canonical(path);
+    std::string s_input = canonicalPath.make_preferred().generic_string();
+    std::regex space("[[:space:]]");
+    std::string s_output;
```
[the remark this answers is not in the document]
Comment (khanaffan) [reply iModelCore/BeSQLite/BeSQLite.cpp]: Fixed

File: iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h [hunk 3]
```diff
@@ -2680,16 +2697,17 @@ struct EXPORT_VTABLE_ATTRIBUTE Db : NonCopyableClass
     };
 
     DbFile* m_dbFile;
-    bool m_isCloudDb = false;
+    bool m_isCloudDb = false; // !URI no need
     StatementCache m_statements;
+    BeDbUri* m_uri; // forward declaration
```
[the remark this answers is not in the document]
Comment (khanaffan) [reply iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h]: `BeDbUri` uses `enums` from `Db:*` e.g. `Db::OpenMode` so we cannot declare `BeDbUri` before `Db` and so I cannot declare a member variable. Compilare complaints about it. There are two solutions
1. Take those Enums out of Db.
2. Use string variable instead  `Utf8String m_uri` and do `Db::GetUri() const { return BeDbUri(m_uri;}`.

For now, I applied 2) solution.

[the remark this answers is not in the document]
Comment (pmconne) [reply iModelCore/BeSQLite/PublicAPI/BeSQLite/BeSQLite.h]: Isn't #1 more straightforward?
```
enum struct DbOpenMode { ... };
struct BeDbUri {
  BeDbUri& SetOpenMode(DbOpenMode v) { ... }
  ...
};
struct Db {
  BeDbUri m_uri;
  using OpenMode = DbOpenMode;
  ...
};
```

**Pull request closed without merging**